### PR TITLE
Add gallery Lucy-Richardson deconvolution algorithm

### LINKED ISSUE #1231: Examples wanted for the gallery
While starting working on the user guide, I noticed that several functions or functionalities are not illustrated in any example of the gallery. Here are a few examples of functionalities that could make nice examples for the gallery (and hence sections of the user guide). Feel free to ask more item to the wishlist in the conversation to follow.
- [x] RGB to HSV conversion, for example for image enhancement (as increasing image saturation), or for image segmentation (in the `data.coffee()` image, a mask of the cup + saucer can be obtained from thresholding the hue channel). WIP: #3607 
-  [x] RGB to gray conversion. Comment on the different weights of the 3 channels, and on the float dtype of the output. 
- [X] gamma and log contrast adjustement. Plot the different kinds of transformations and show the result when applied to a grayscale image such as `moon` DONE: #1670
- [X] Richardson-Lucy deconvolution. DONE: #2376 
- [ ] LPIFilter2D / inverse filters (I have no idea how they are used. @stefanv ?)
- [ ] ``measure.block_reduce``
- [ ] ``measure.profile_line``

### PR DESCRIPTION
**Pull request by souravsingh**
Adds a deconvolution example based on Lucy-Richardson deconvolution algorithm.

Part of #1231

### CODE REVIEW

File: doc/examples/filters/plot_deconvolution.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+"""
+=====================
+Image Deconvolution
+=====================
+In this example, we deconvolve an image using Richardson-Lucy
+deconvolution algorithm.
+Richardson-Lucy deconvolution
+-----------------------------
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: I would say, remove lines 8 and 9 here — an intermediate heading doesn't add much to this small example.

File: doc/examples/filters/plot_deconvolution.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+The algorithm is based on Bayesian-based iterative method
+where an image blurred by a point-spread function is 
+recovered. 
+.. [1] William Hadley Richardson, "Bayesian-Based Iterative
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: Add a blank line above this.

File: doc/examples/filters/plot_deconvolution.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+.. [1] William Hadley Richardson, "Bayesian-Based Iterative 
+       Method of Image Restoration",
+       J. Opt. Soc. Am. A 27, 1593-1607 (1972)
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage import color, data, restoration
+
+from scipy.signal import convolve2d as conv2
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: I would prefer here:

```python
from scipy import ndimage as ndi

# [...]

astro_blurred = ndi.convolve(astro, psf, 'same')
```

File: doc/examples/filters/plot_deconvolution.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+astro = color.rgb2gray(data.astro())
+psf = np.ones((5, 5)) / 25
+astro = conv2(astro, psf, 'same')
+camera += 0.1 * astro.std() * np.random.standard_normal(astro.shape)
+
+deconvolved = restoration.richardson_lucy(astro, psf, 5)
+
+fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(8, 5),
+                       sharex=True, sharey=True,
+                       subplot_kw={'adjustable': 'box-forced'})
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: What do you think about including the original image in the plot as well? I think it would help people to see where we are starting from. Then I would use the titles, "Original data", "Blurred", "Deconvolved using Richardson-Lucy".

File: doc/examples/filters/plot_deconvolution.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+ax[0].imshow(astro, vmin=deconvolved.min(), vmax=deconvolved.max())
+ax[0].axis('off')
+ax[0].set_title('Data')
+
+ax[1].imshow(deconvolved)
+ax[1].axis('off')
+ax[1].set_title('Deconvolution using Lucy-Richardson')
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: Lucy-Richardson -> Richardson-Lucy.

File: doc/examples/filters/plot_deconvolution.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+In this example, we deconvolve an image using Richardson-Lucy
+deconvolution algorithm.
+Richardson-Lucy deconvolution
+-----------------------------
+The algorithm is based on Bayesian-based iterative method
+where an image blurred by a point-spread function is 
+recovered. 
+.. [1] William Hadley Richardson, "Bayesian-Based Iterative 
+       Method of Image Restoration",
+       J. Opt. Soc. Am. A 27, 1593-1607 (1972)
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Please add DOI is available. `DOI:xxxx.xxx.xxx`

Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Please, add a reference to wikipedia for the reader who doesn't know the details: https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution

**Comment by souravsingh** [comment]
Thanks for the review @sciunto and @jni

File: doc/examples/filters/plot_deconvolution.py (hunk 3 above)
Comment (tseclaudia) [inline doc/examples/filters/plot_deconvolution.py]: Delete blank lines 19 and 21.

**Comment by souravsingh** [comment]
Is there any more changes that are needed?

File: doc/examples/filters/plot_deconvolution.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+.. [2] http://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage import color, data, restoration
+
+from scipy import ndimage as ndi
+
+astro = color.rgb2gray(data.astro())
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: I guess so:
```
Traceback (most recent call last):
  File "doc/examples/filters/plot_deconvolution.py", line 26, in <module>
    astro = color.rgb2gray(data.astro())
AttributeError: 'module' object has no attribute 'astro'
```

**Comment by souravsingh** [comment]
Richardson-Lucy example was done before here-https://github.com/scikit-image/scikit-image/pull/1685
Do I update my PR according to the previous one?

File: doc/examples/filters/plot_deconvolution.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: Remove this line.

File: doc/examples/filters/plot_deconvolution.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+"""
+=====================
+Image Deconvolution
+=====================
+In this example, we deconvolve an image using Richardson-Lucy
+deconvolution algorithm.
+
+The algorithm is based on Bayesian-based iterative method
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: "based, ...-based" :grimacing: Maybe rephrase this somehow?

File: doc/examples/filters/plot_deconvolution.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage import color, data, restoration
+
+from scipy import ndimage as ndi
+
+astro = color.rgb2gray(data.astronaut())
+psf = np.ones((5, 5)) / 25
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: `np.full`?

File: doc/examples/filters/plot_deconvolution.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+astro_blurred = ndi.convolve(astro, psf)
+astro += 0.1 * astro.std() * np.random.standard_normal(astro.shape)
+
+deconvolved = restoration.richardson_lucy(astro, psf, 5)
+
+fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(8, 5), sharex=True, sharey=True)
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: Please, wrap this line (it is longer than 80 chars).

File: doc/examples/filters/plot_deconvolution.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+ax1.imshow(astro)
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: In my environment all imshow calls plot in non-grayscale colormap by default.
You should add `, cmap=plt.cm.gray)` or any other workaround.

File: doc/examples/filters/plot_deconvolution.py (hunk 11 above)
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: I think it is better to visualize the results in 2x2 grid of plots.
Here is how it looks at the moment:
default.

Also, please, adjust `figsize` to increase the effective area.

File: doc/examples/filters/plot_deconvolution.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+ax2.imshow(astro_blurred)
+ax2.set_title('Blurred Image')
+ax2.axis('off')
+ax2.set_adjustable('box-forced')
+
+ax3.imshow(deconvolved)
+ax3.set_title('Deconvolved using Richardson-Lucy')
+ax3.axis('off')
+ax3.set_adjustable('box-forced')
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: We decided to move these configuration lines into a single `for` loop to reduce visual noise. See https://github.com/soupault/scikit-image/blob/7018540c9073593bc51f35760ad769ff1f329b81/doc/examples/xx_applications/plot_coins_segmentation.py#L43-L45.

File: doc/examples/filters/plot_deconvolution.py (hunk 11 above)
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: BTW, I'm personally satisfied with the results. Could we adjust some parameters of algorithm to increase the quality of the output?

File: doc/examples/filters/plot_deconvolution.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ax1.set_title('Original image')
+
+ax2.imshow(astro_blurred, cmap=plt.cm.gray)
+ax2.set_title('Blurred Image')
+
+ax3.imshow(deconvolved, cmap=plt.cm.gray)
+ax3.set_title('Deconvolved using Richardson-Lucy')
+
+for a in axis:
+       a.axis('off')
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: four spaces as indentation, following the PEP8 Python style convention.

File: doc/examples/filters/plot_deconvolution.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+ax1.imshow(astro, cmap=plt.cm.gray)
+ax1.set_title('Original image')
+
+ax2.imshow(astro_blurred, cmap=plt.cm.gray)
+ax2.set_title('Blurred Image')
+
+ax3.imshow(deconvolved, cmap=plt.cm.gray)
+ax3.set_title('Deconvolved using Richardson-Lucy')
+
+for a in axis:
```
Comment (jni) [inline doc/examples/filters/plot_deconvolution.py]: `axis` is not a variable. `for a in (ax1, ax2, ax3):`

**Comment by jni** [comment]
@souravsingh looks like this needs one more round of changes.

**Comment by souravsingh** [comment]
@jni I have made the changes in accordance to the review.

**Comment by jni** [comment]
All good from me! Thanks @souravsingh!

**Comment by jni** [comment]
@sciunto @soupault can you finalise your reviews and merge? Thanks!

**Comment by soupault** [comment]
@jni @souravsingh I still have two requests pending:
https://github.com/scikit-image/scikit-image/pull/2376#discussion_r88618827

image
image

File: doc/examples/filters/plot_deconvolution.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
+"""
+=====================
+Image Deconvolution
+=====================
+In this example, we deconvolve an image using Richardson-Lucy
+deconvolution algorithm.
+
+The algorithm is based on a PSF(Point Spread Function).
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Missing space before (

File: doc/examples/filters/plot_deconvolution.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+The blurred image is sharpened  through a number of 
+iterations, which needs to be hand-tuned.
+
+.. [1] William Hadley Richardson, "Bayesian-Based Iterative 
+       Method of Image Restoration",
+       J. Opt. Soc. Am. A 27, 1593-1607 (1972), doi:10.1364
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Please, fix the DOI as `DOI:10.1364/JOSA.62.000055`

File: doc/examples/filters/plot_deconvolution.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
+"""
+=====================
+Image Deconvolution
+=====================
+In this example, we deconvolve an image using Richardson-Lucy
+deconvolution algorithm.
+
+The algorithm is based on a PSF(Point Spread Function).
+The blurred image is sharpened  through a number of
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: remove 1 space before though

File: doc/examples/filters/plot_deconvolution.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage import color, data, restoration
+
+astro = color.rgb2gray(data.astronaut())
+
+from scipy.signal import convolve2d as conv2
+psf = np.ones((5, 5))/25
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: spaces around /

File: doc/examples/filters/plot_deconvolution.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+astro = conv2(astro, psf, 'same')
+astro_noisy = astro.copy()
+astro_noisy += (np.random.poisson(lam=25, size=astro.shape)-10)/255.
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: spaces around /

File: doc/examples/filters/plot_deconvolution.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+.. [2] http://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+
+from skimage import color, data, restoration
+
+astro = color.rgb2gray(data.astronaut())
+
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Please, add a comment to say that you are artificially degrading the image.

File: doc/examples/filters/plot_deconvolution.py [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+from scipy.signal import convolve2d as conv2
+psf = np.ones((5, 5))/25
+astro = conv2(astro, psf, 'same')
+astro_noisy = astro.copy()
+astro_noisy += (np.random.poisson(lam=25, size=astro.shape)-10)/255.
+
+deconvolved_RL = restoration.richardson_lucy(astro_pois, psf, iterations=30)
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Please, add a comment to say that your restore the image

File: doc/examples/filters/plot_deconvolution.py (hunk 20 above)
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: spaces around -

File: doc/examples/filters/plot_deconvolution.py (hunk 22 above)
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: `astro_pois` looks undefined.

File: doc/examples/filters/plot_deconvolution.py (hunk 16 above)
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Can you elaborate about PSF, please?

File: doc/examples/filters/plot_deconvolution.py [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+astro_noisy = astro.copy()
+astro_noisy += (np.random.poisson(lam=25, size=astro.shape) - 10) / 255.
+
+# Restore Image using Richardson-Lucy algorithm
+deconvolved_RL = restoration.richardson_lucy(astro_noisy, psf, iterations=30)
+
+fig, ax = plt.subplots(nrows=1, ncols=3, figsize=(8, 5))
+plt.gray()
+
+for a in (ax1, ax2, ax3):
```
Comment (soupault) [inline doc/examples/filters/plot_deconvolution.py]: Where do `ax1, ...` come from?

Comment (souravsingh) [inline doc/examples/filters/plot_deconvolution.py]: My bad! I will fix it

**Comment by souravsingh** [comment]
@sciunto Is the Example ready to merge?

File: doc/examples/filters/plot_deconvolution.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
+"""
+=====================
+Image Deconvolution
+=====================
+In this example, we deconvolve an image using Richardson-Lucy
+deconvolution algorithm.
+
+The algorithm is based on a PSF (Point Spread Function),
+where PSF is described as the impulse response of the 
+optical system .
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: remove space before .

File: doc/examples/filters/plot_deconvolution.py [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+The blurred image is sharpened through a number of
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: no new line please.

File: doc/examples/filters/plot_deconvolution.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+       a.axis('off')
+
+ax[0].imshow(astro)
+ax[0].set_title('Original Data')
+
+ax[1].imshow(astro_noisy)
+ax[1].set_title('Noisy data')
+
+ax[2].imshow(deconvolved_RL, vmin=astro_noisy.min(), vmax=astro_noisy.max())
+ax[2].set_title('Restoration using\nRichardson Lucy')
```
Comment (sciunto) [inline doc/examples/filters/plot_deconvolution.py]: Richardson-Lucy

**Comment by sciunto** [comment]
@souravsingh Do you need help to squash the commits?

**Comment by souravsingh** [comment]
@sciunto Is the PR ready for merging?

**Comment by sciunto** [comment]
Perfect! I merge it. Many thanks to you!

**Comment by souravsingh** [comment]
@sciunto Sorry for asking this late. The Pull Request #1685 which implements Richardson-Lucy Example Gallery is open. I suppose we should close it?

**Comment by sciunto** [comment]
yes, thanks.
